src/UpgradeRule/PHP/Visitor/ClassWarningsVisitor.php

```diff
@@ -0,0 +1,76 @@
+<?php
+
+namespace SilverStripe\Upgrader\UpgradeRule\PHP\Visitor;
+
+use PhpParser\Node;
+use PhpParser\Node\Stmt\Class_;
+use PhpParser\Node\Expr\New_;
+use PhpParser\Node\Expr\StaticCall;
+use SilverStripe\Upgrader\Util\Warning;
+use SilverStripe\Upgrader\Util\ApiChangeWarningSpec;
+
+/**
+ * Relies on {@link SymbolContextVisitor} to annotate the
+ * 'symbolContext' attribute of nodes.
+ *
+ * @package SilverStripe\Upgrader\UpgradeRule
+ */
+class ClassWarningsVisitor extends WarningsVisitor
+{
+
+    public function enterNode(Node $node)
+    {
+        parent::enterNode($node);
+
+        $isClassNode = (
+            $node instanceof Class_ ||
+            $node instanceof StaticCall ||
+            $node instanceof New_
+        );
+        if ($isClassNode) {
+            foreach ($this->specs as $spec) {
+                if (!$this->matchesSpec($node, $spec)) {
+                    continue;
+                }
+
+                $this->addWarning($node, $spec);
+            }
+        }
+    }
+
+    /**
+     * @param Node $node
+     * @param ApiChangeWarningSpec $spec
+     * @return bool
+     */
+    protected function matchesSpec(Node $node, ApiChangeWarningSpec $spec)
+    {
+        $symbol = $spec->getSymbol();
+        $context = $node->getAttribute('symbolContext');
+
+        $class = '';
+
+        if (isset($node->name)) {
+            $class = $node->name;
+        }
+
+        // extends MyNamespace\MyClass or extends MyClass
+        if (isset($node->extends->parts)) {
+            $class = implode('\\', $node->extends->parts);
```

Comment:
If this class matches ($node->name) but the parent doesn't ($node->extends->parts), then $class gets overwritten with the non-matching parent, and the check fails when it should succeed.

A fix would be to check `$class === $symbol` after each check instead of once at the end.

Reply:
Good catch! Done